src/include/concurrency/decentralized_epoch_manager.h

```diff
@@ -0,0 +1,303 @@
+//===----------------------------------------------------------------------===//
+//
+//                         Peloton
+//
+// decentralized_epoch_manager.h
+//
+// Identification: src/include/concurrency/decentralized_epoch_manager.h
+//
+// Copyright (c) 2015-16, Carnegie Mellon University Database Group
+//
+//===----------------------------------------------------------------------===//
+
+
+#pragma once
+
+#include <thread>
+#include <vector>
+
+#include "common/macros.h"
+#include "type/types.h"
+#include "common/logger.h"
+#include "common/platform.h"
+#include "common/init.h"
+#include "common/thread_pool.h"
+#include "concurrency/epoch_manager.h"
+
+namespace peloton {
+namespace concurrency {
+
+
+class DecentralizedEpochManager : public EpochManager {
+  DecentralizedEpochManager(const DecentralizedEpochManager&) = delete;
+
+struct Epoch {
+  std::atomic<size_t> read_only_count_;
+  std::atomic<size_t> read_write_count_;
+
+  Epoch(): 
+    read_only_count_(0), 
+    read_write_count_(0) {}
+
+  Epoch(const Epoch &epoch): 
+    read_only_count_(epoch.read_only_count_.load()), 
+    read_write_count_(epoch.read_write_count_.load()) {}
+
+  void Init() {
+    read_only_count_ = 0;
+    read_write_count_ = 0;
+  }
+};
+
+struct LocalEpochContext {
+
```

Comment:
For better cache performance, we need to force the LocalEpochContext to take up a cache line. I think for now sizeof(LocalEpochContext) is greater than 64 bytes. But I suggest that either we enforce it in the code or leave a comment about it.